File: 2023/Day4.cs
using System.Text.RegularExpressions;

class Day4
{
    static void Main()
    {
        string[] data = File.ReadLines("input.txt").ToArray();
        Console.WriteLine("Part 1 Answer: " + Part1(data));
        Console.WriteLine("Part 2 Answer: " + Part2(data));
    }

    static int Part1(string[] data)
    {
        int total = 0;
        for (int i = 0; i < data.Length; i++)
        {
            string[] line = data[i].Substring(7 + (i + 1).ToString().Length).Split(" | ");
            string[] winningCards = Regex.Replace(line[0].Trim(' '), @"\s+", " ").Split(" ");
            string[] haveCards = Regex.Replace(line[1].Trim(' '), @"\s+", " ").Split(" ");
            IEnumerable<string> haveWinners = winningCards.Intersect(haveCards);
            total += (int)Math.Pow(2, haveWinners.Count<string>() - 1);
        }
        return total;
    }

    static int Part2(string[] data)
    {
        int total = 0;
        int[] cardCopies = new int[data.Length];
        for (int i = 0; i < cardCopies.Length; i++)
        {
            cardCopies[i] = 1;
        }
        for (int i = 0; i < data.Length; i++)
        {
            string[] line = data[i].Substring(7 + (i + 1).ToString().Length).Split(" | ");
            string[] winningCards = Regex.Replace(line[0].Trim(' '), @"\s+", " ").Split(" ");
            string[] haveCards = Regex.Replace(line[1].Trim(' '), @"\s+", " ").Split(" ");
            IEnumerable<string> haveWinners = winningCards.Intersect(haveCards);
            for (int j = i + 1; j < i + 1 + haveWinners.Count<string>(); j++)
            {
                cardCopies[j] += cardCopies[i];
            }
            total += cardCopies[i];
        }
        return total;
    }
}

File: 2024/Day4.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

int part1(const std::vector<std::string> &input)
{
    int count = 0;
    for (int i = 0; i < input.size(); i++)
    {
        for (int j = 0; j < input[i].size(); j++)
        {
            if (input[i][j] == 'X')
            {
                for (int x = -1; x <= 1; x++)
                {
                    for (int y = -1; y <= 1; y++)
                    {
                        if (x == 0 && y == 0)
                        {
                            continue;
                        }
                        if (i + 3 * x < 0 || i + 3 * x >= input.size() || j + 3 * y < 0 || j + 3 * y >= input[i].size())
                        {
                            continue;
                        }
                        if (input[i + x][j + y] == 'M' && input[i + 2 * x][j + 2 * y] == 'A' && input[i + 3 * x][j + 3 * y] == 'S')
                        {
                            count++;
                        }
                    }
                }
            }
        }
    }
    return count;
}

int part2(const std::vector<std::string> &input)
{
    int count = 0;
    for (int i = 1; i < input.size() - 1; i++)
    {
        for (int j = 1; j < input[i].size() - 1; j++)
        {
            if (input[i][j] == 'A')
            {
                if (input[i + 1][j + 1] == 'M' && input[i - 1][j - 1] == 'S' || input[i + 1][j + 1] == 'S' && input[i - 1][j - 1] == 'M')
                {
                    if (input[i + 1][j - 1] == 'M' && input[i - 1][j + 1] == 'S' || input[i + 1][j - 1] == 'S' && input[i - 1][j + 1] == 'M')
                    {
                        count++;
                    }
                }
            }
        }
    }
    return count;
}

int main()
{
    std::ifstream file("input.txt");
    std::vector<std::string> input;
    std::string line;
    while (std::getline(file, line))
    {
        input.push_back(line);
    }
    file.close();

    fprintf(stdout, "Part 1 Answer: %d\n", part1(input));
    fprintf(stdout, "Part 2 Answer: %d\n", part2(input));

    return 0;
}